### PR TITLE
feat: Add WithResourceOption for additional resource configuration

### LINKED ISSUE #12: Support adding custom resource.Detectors / resource.Options
I'd like to be able to add custom Detectors, like the ones [in sdk/resource](https://pkg.go.dev/go.opentelemetry.io/otel/sdk/resource#Option) or the [ones in contrib](https://github.com/open-telemetry/opentelemetry-go-contrib/tree/main/detectors) for cloud providers.

There are two problems here:

1) I _think_ the fix is something like adding `func WithResourceOptions(...resource.Option) Option` and then passing those to the `resource.New` call [here](https://github.com/honeycombio/otel-launcher-go/blob/main/launcher/launcher.go#L409-L413) (and `sdk/resource.WithDetectors` returns a `resource.Option` so that covers the contrib Detectors too). However, the error returned from `resource.New` is ignored right now and I know at least some of the AWS Detectors will return errors when not on an appropriate platform (eg, the EKS detector returns and error when not running on EKS). I'm not sure if it's worth threading that error all the way back up to the caller of `ConfigureOpenTelemetry` or if there are other options I'm not seeing. I'd like some advice before attempting a fix.

2) The `semconv` versions imported in this project ([v1.10.0](https://github.com/honeycombio/otel-launcher-go/blob/main/launcher/launcher.go#L22)) and in those other libraries ([v1.17.0](https://github.com/open-telemetry/opentelemetry-go-contrib/blob/main/detectors/aws/ec2/ec2.go#L28)) are different, meaning that the `semconv.SchemaURL` values are different, meaning that any attempt to merge them together will fail. Happy to submit a PR to upgrade this package's version if that sounds like a reasonable fix.

### PR DESCRIPTION
**Pull request by robbkidd**
## Which problem is this PR solving?

We are unable to add resource detectors when configuring the launcher.

- Closes #12

## Short description of the changes

Implementation starts with what was written for #47 with smoke tests added to verify the order of precedence for resources attributes from environment over code.

[vera] upon further inspection, the original behavior of CODE attributes winning over ENV attributes was valid:

> The SDK MUST extract information from the OTEL_RESOURCE_ATTRIBUTES environment variable and [merge](https://opentelemetry.io/docs/specs/otel/resource/sdk/#merge) this, as the secondary resource, with any resource information provided by the user, i.e. the user provided resource information has higher priority.
— [OTel Resource SDK spec](https://opentelemetry.io/docs/specs/otel/resource/sdk/#specifying-resource-information-via-an-environment-variable)

“user provided resource information” -- the user could mean the application code writer OR the operator setting environment variables. It’s ambiguous! And *not* settled in the spec: https://github.com/open-telemetry/opentelemetry-specification/issues/1620

See note in that issue: 
>others in the Go SIG have interpreted this passage to require that resource attributes from the environment must be subordinated to any resource attributes provided by the developer-user

This is how the underlying Go SDK works, as designed.

The CODE>ENV decision was also made in the Ruby SDK: https://github.com/robertlaurin/opentelemetry-ruby/blob/0c03e4aee09093ce0d80142e13a35326aad9f877/sdk/test/opentelemetry/sdk/configurator_test.rb#L35

Description from original PR:

Rather than adding a specific `WithDetectors` func to mirror the functionality in the [resource package](https://pkg.go.dev/go.opentelemetry.io/otel/sdk/resource#WithDetectors) this adds a more generic version so that any resource option configuration can be provided.

This has been done in a backwards compatible manner but I could see a future where the `WithResourceAttributes` function is removed as it is duplicate functionality now.

## How to verify that this has the expected result

- [x] Unit tests - logic's probably sound
- [x] Smoke tests - interaction with config from environment behaves according to spec

### CODE REVIEW

File: examples/main.go [hunk 1]
```diff
@@ -5,20 +5,28 @@ import (
...
 
 	if err != nil {
 		log.Fatalf("error setting up OTel SDK - %e", err)
 	}
 
 	defer otelShutdown()
 	tracer := otel.Tracer("my-app")
-	ctx := context.Background()
-	ctx, span := tracer.Start(ctx, "doing-things")
+	_, span := tracer.Start(context.Background(), "doing-things")
```
Comment (robbkidd) [inline examples/main.go:30]: I … don't know why I included this. It has nothing to do with the resource attributes issue. 😐

**Comment by robbkidd** [comment]
## What's up with the revert/readd of this feature?

Resource attributes set via environment variable must supercede resource attributes set in code.

> The SDK MUST extract information from the OTEL_RESOURCE_ATTRIBUTES environment variable and [merge](https://opentelemetry.io/docs/specs/otel/resource/sdk/#merge) this, as the secondary resource, with any resource information provided by the user, i.e. the user provided resource information has higher priority.
— [OTel Resource SDK spec](https://opentelemetry.io/docs/specs/otel/resource/sdk/#specifying-resource-information-via-an-environment-variable)

#47 introduced `WithResource()` but the current order of merging resource attributes has [code-provided attributes](https://github.com/honeycombio/otel-config-go/blob/6365f2db2824c5a112dfdad5f9a215a0b83790ff/otelconfig/otelconfig.go#L360-L362) (from this function) clobbering [user-provided attributes from the environment](https://github.com/honeycombio/otel-config-go/blob/6365f2db2824c5a112dfdad5f9a215a0b83790ff/otelconfig/otelconfig.go#L344). It's important to remember that the "user" mentioned by the spec is not the developer writing application code, it's the person using the application later at runtime.

File: otelconfig/otelconfig.go [hunk 2]
```diff
@@ -411,11 +420,17 @@ func newResource(c *Config) *resource.Resource {
...
 
+	baseOptions := []resource.Option{
+		resource.WithSchemaURL(semconv.SchemaURL),
+		resource.WithAttributes(attributes...),
+	}
+
+	options := append(baseOptions, c.ResourceOptions...)
+
 	// These detectors can't actually fail, ignoring the error.
 	r, _ = resource.New(
```
Comment (dstrelau) [inline otelconfig/otelconfig.go:431]: It seems like this comment is no longer true, as there are arbitrary Options here now, right? Does this need to do something with the error?

**Comment by vreynolds** [comment]
Updated the description with my reasoning for walking back on the premise of the revert. I believe code config wins over env config, at least in the Go SDK land.

**Review by JamieDanielson — APPROVED** [review]
It seems like we're still doing some extra work here that could be simplified by using the available resource options now available upstream, but that can be addressed separately. This at least allows the end user more flexibility in setting additional resource options. Appreciate the PR details and the test cleanup as well 👍

**Pull request merged**